AI Review updates for Example PSE & PSE 1 (#57)
Ada-Developers-Academy/core-pses: about-pses/example-pse.md

<br>

In the box below, please submit:
1. A shareable link to your conversation in ChatGPT
    - [Documentation for creating a shareable link in ChatGPT](https://help.openai.com/en/articles/7925741-chatgpt-shared-links-faq)
2. Your reflections and summary of the discussion with ChatGPT

##### !end-question
##### !hint

**Troubleshooting**
- If you are having issues with the tool understanding the prompt, try formatting the problem statement or examples differently.
- If you’ve reformatted the information and are still not getting useful results, reach out in #study-hall and share what you are experiencing and the link to your chat so folks can take a look and help you troubleshoot!

<br>

**Summarizing the Review**
- Did the AI tool uncover anything about the observations you made that you hadn’t considered?
- Did the AI tool uncover anything about the questions you asked that you hadn’t considered?
- Did the AI tool suggest updates to the observations you made or questions you asked? 
    - If so, what updates and why?
- Did the AI tool suggest any new observations or questions?
    - If so, what? Why would they be useful?

##### !end-hint
##### !explanation 

> Everyone’s observations & questions, and conversation with ChatGPT will be a little different, thus, the summaries will look a little different. As an example, let’s say we used the sample response from "Describe Your Understanding" above to fill in the prompt template.

Author: ```suggestion
Everyone’s observations & questions, as well as their conversations with ChatGPT. will be a little different. Because of this, the summaries created will look a little different.

After you submit a response in Learn, you'll see "Show Explanation" appear and when it is clicked an explanation is revealed. For this example, we'll use the questions and explanations from the "Explanation" section of the previous question as examples to create our prompt. 
```

## Prompts

<!-- prettier-ignore-start -->
### !challenge
* type: paragraph
* id: 6957fd47-9dba-48b9-ac9c-38315cda0b28
* title: Describe Your Understanding
* topics: pse
##### !question

Before you begin solving this problem, take a moment to think like a professional software engineer. 
- What do we know about the problem? 
- What assumptions can we make based on the information in the problem statement? 
- What further information do the example inputs and outputs give us?
- What questions would you ask a teammate, product manager, or interviewer to better understand the problem before writing any code?

<br>

In the box below, list 5 or more observations about the problem or questions whose answers would clarify the problem statement. For each observation or question, include information on why that observation is important or why you are asking the question
- For each observation, answer how that observation will affect your approach to the problem
- For each question, describe what you are hoping to clarify about the problem and provide an answer which includes the effect your decision would have on how you might approach the problem.

<br>

> As you come up with observations and questions, assume that error handling for invalid data is managed outside the function. We want to focus on the core behavior of the function we will write. 

Reviewer: Repeat from comment on `highest-rated.md`

```suggestion
As you come up with observations and questions, assume that error handling for invalid data is managed outside the function. We want to focus on the core behavior of the function we will write. Examples of invalid data we don't need to manage may include:
- Mismatched data types: If a function expects a string, you may assume it will receive a string. This follows for nested data as well.
- String Sanitization: You may assume strings will not include invalid characters.
```

##### !end-question
##### !hint

Further questions to ask as you read through the problem statement and examples:
- What is the goal of the function?
- What are the types of the expected inputs and outputs?
- Are there any restrictions on any of the inputs?
  - For example: if any of the inputs are a list, do we know anything about how the list is ordered?
- What do the examples show us about the data types and values that are allowed for our inputs?
- What do the examples tell us about the return value in different scenarios?

> - Reflect on the observations you have made so far, what questions would give you new information?

Reviewer: Repeat comment from `highest-rated.md`

```suggestion
- Reflecting on the observations you have made so far, what questions would give you new information?
```

> - Assume error handling for invalid data is managed outside the function. 

Reviewer: Repeated comment from `highest-rated.md`
```suggestion
```

<br>

We should also avoid questions that change the observable behavior of a prompt. While we shouldn't ask about changing parameters or return values, or modifying behavior to something that conflicts with the prompt, we could observe the requirements of the prompt, and mention how we might approach it if it were solely up to us. 

<br>

In an interview setting, this can give a sense of our thought process, and convey familiarity with the language over all, opening up space for a discussion. Ideally, we should think about observations and questions that would help us understand the problem better and make decisions about how to approach it.

<br>

One of many possible sample responses could look like:

<br>

1. I need to implement a function called `winner` that takes two arguments: `player_1` and `player_2`.
    - This means I need to create a function definition with two parameters named exactly `player_1` and `player_2`.

> 2. Each argument represents a move for player_1 or player_2 and will be one of "rock", "paper", or "scissors".

Reviewer: ```suggestion
2. Each argument represents a move for `player_1` or `player_2` and will be one of "rock", "paper", or "scissors".
```

    - Aside from checking if someone wins, I need to consider the edge case where the players could tie.

4. The second input/output example shows both players choosing "rock" and the result is a tie
    - I am hypothesizing that two players choosing the same move results in a tie. 
    - This is supported by the table of rules further below that shows the result as "It's a tie!" any time both players choose the same move.

5. The table of rules contains text "input (first argument)" and "input (second argument)"
    - This tells me when deciding who is the winner, I need to be mindful that the first argument's value always represents player_1 and the second argument represents player_2.
    - This is supported by the first example where `player_1 = "rock"` and `player_2 = "scissors"`, and the result was `"Player 1 wins!"`.

6. Based on the table of rules: "rock" beats "scissors", "scissors" beats "paper", "paper" beats "rock".
    - When developing my solution, I will use these rules to guide the conditional statements I need to write to determine the winner.

<br>

> We should keep in mind that sometimes, areas of clarification may not jump out at us when we first read the prompt. It's completely OK to think through the prompt, and even start working on it. As we start to write code, we may note something we missed before or realize that we need to ask a clarifying question. This is a normal part of the problem-solving process. So don't feel like we need to note everything or ask all our clarifying questions before we start working on the problem.

Reviewer: ```suggestion
We should keep in mind that sometimes, areas of clarification may not jump out at us when we first read the prompt. It's completely OK to think through the prompt, and even start working on it. As we start to write code, we may note something we missed before or realize that we need to ask a clarifying question. This is a normal part of the problem-solving process, so don't feel like we need to note everything or ask all our clarifying questions before we start working on the problem.
```

Reviewer 2: This explanation is a rather long block of text and I feel like this is important to visually highlight this point. How do you feel about italicizing or bolding it?

```suggestion
We should keep in mind that sometimes, areas of clarification may not jump out at us when we first read the prompt. It's completely OK to think through the prompt, and even start working on it. **As we start to write code, we may note something we missed before or realize that we need to ask a clarifying question. This is a normal part of the problem-solving process. So don't feel like we need to note everything or ask all our clarifying questions before we start working on the problem.**
```

Author: To help break up the chunk of text, how do y'all feel about moving that last couple sentences to a bullet point? On re-reading, I also didn't love telling people "Don't feel a certain way!", and rephrased it a little. 

```suggestion
We should keep in mind that sometimes, areas of clarification may not jump out at us when we first read the prompt. It's completely OK to think through the prompt, and even start working on it. As we start to write code, we may note something we missed before or realize that we need to ask a clarifying question. 
- This is a normal part of the problem-solving process! We don't need to note everything or ask all our clarifying questions before we start working on the problem.
```

Reviewer 2: Making it a bullet point looks good to me!

### !end-challenge
<!-- prettier-ignore-end -->

<!-- prettier-ignore-start -->
### !challenge
* type: paragraph
* id: d14c0645-1681-4874-ba8c-37e1bb19262d
* title: Review Observations & Questions
* topics: pse
##### !question

While we build our skills in breaking down a problem and choosing clarifying questions, let’s use an external tool like ChatGPT to review the observations and questions we wrote while describing our understanding. 

<br>

Our goals are to: 
- confirm if our observations and assumptions make sense in the context of the code problem
- ensure we are asking questions that will tell us new information about the problem space
- check our understanding of the information we expect to get from those questions
- uncover other observations that would help shape our approach and understand how they would affect our approach
- uncover further questions that could be useful to ask and understand why those other questions could be helpful

<br>

For this question we will:
1. Build a prompt using [the template linked here](https://gist.githubusercontent.com/ada-instructors/16c97dc4b16ab2bf449d9d7a81caeb16/raw/pse_observations_questions_review_template.md)

> 2. Share the filled in prompt with an AI tool like ChatGPT

Reviewer: Repeated comment from `highest-rated.md`:

```suggestion
2. Share the completed prompt with an AI tool like ChatGPT
```

> 3. After the initial review, ask *at least one* follow up question using the AI tool that furthers your understanding of the problem and why certain observations or questions are useful. Some examples could be asking questions to: 

Reviewer: Repeated comment from `highest-rated.md`:
```suggestion
3. After the initial review, ask the AI tool *at least one* follow up question that furthers your understanding of the problem and why certain observations or questions are useful. Some examples could be asking questions to: 
```

## Prompts

<!-- prettier-ignore-start -->
### !challenge
* type: paragraph
* id: 6957fd47-9dba-48b9-ac9c-38315cda0b28
* title: Gather Information
* topics: pse
##### !question

Use the box below to describe your understanding of the problem. Format the information in whatever way makes sense for you to organize your thoughts and review the contents as you continue with the upcoming prompts. Consider:
- What do we know about the problem? 
- What assumptions can we make based on the information in the problem statement? 
- What further information do the example inputs and outputs give us?

##### !end-question
##### !hint

Further questions to ask as you read through the problem statement and examples:
- What is the goal of the function?
- What are the types of the expected inputs and outputs?

> - Are there any restrictions on any of the inputs?

Author: This has been added, resolving before new reviews

> As stated, this list contains several questions that overlap in what they are asking as an example of different perspectives. Try not to include multiple questions about the same topic in your own questions. For example, if we ask about how to handle invalid input, we only need to ask once, not in three different ways as we have here, especially since the sample answers here are contradictory!

Author: Were you thinking something like italics, or breaking up the paragraph some? A couple options:

```suggestion
As stated, this list contains several questions that overlap in what they are asking as an example of different perspectives. _Try not to include multiple questions about the same topic in your own questions._ For example, if we ask about how to handle invalid input, we only need to ask once, not in three different ways as we have here, especially since the sample answers here are contradictory!
```

```suggestion
As stated, this list contains several questions that overlap in what they are asking as an example of different perspectives. Try not to include multiple questions about the same topic in your own questions. 
- For example, if we ask about how to handle invalid input, we only need to ask once, not in three different ways as we have here, especially since the sample answers here are contradictory!
```

Reviewer 2: I think breaking it into chunks and using the bolding could be helpful because visually it is a long block of text.

> As stated, this list contains several questions that overlap in what they are asking as an example of different perspectives. Try not to include multiple questions about the same topic in your own questions. For example, if we ask about how to handle invalid input, we only need to ask once, not in three different ways as we have here, especially since the sample answers here are contradictory!

Author: ```suggestion
As stated, this list contains several questions that overlap in what they are asking as an example of different perspectives. *Try not to include multiple questions about the same topic in your own questions.*
- For example, if we ask about how to handle invalid input, we only need to ask once, not in three different ways as we have here, especially since the sample answers here are contradictory!
```

    - Since the output examples do not mention anything printed out, I will assume that we are only returning the result as a string
5. **I'm familiar with a variant of rock, paper, scissors that includes additional throws (lizard and Spock). Should I keep my approach focused on the basic rules, or should I use an implementation that could be extended to game variants?**
    - I want to know if my implementation needs to be flexible enough to extend it with additional game rules.
    - While I could write a helper function that accepts a data structure that encodes the valid moves and what beats what, I'll assume that I should focus on the basic rules for now. This way, I can get a working solution faster, and if I need to extend the game later, I can refactor my code to be more flexible.
6. **When friends use rock, paper, scissors to make decisions, there are often more than two players. Should I keep my approach focused on the basic rules for 2 players?**
    - I want to know if I should I use an implementation that could be extended to support multiple players simultaneously.

>     - While I could write a helper function that accepts a list of moves made by all players, the multiplayer version introduces a few complications into resolving a round that I'd like to avoid for now. I'll assume that I should focus on the basic rules for now. This way, I can get a working solution faster, and if I need to extend the game later, I can refactor my code to be more flexible.

Author: I like that, it feels a little less vague and focuses in on where we want students putting effort. How does this sound to incorporate that language?

```suggestion
    - While I could write a helper function that accepts a list of moves made by all players, the multiplayer version increases the scope of the problem. I'll assume that I should focus on the basic rules for now. This way, I can get a working solution faster, and if I need to extend the game later, I can refactor my code to be more flexible.
```

Reviewer 2: Yeah! That looks good to me.

### !end-challenge
<!-- prettier-ignore-end -->

<!-- prettier-ignore-start -->
### !challenge
* type: paragraph
* id: d14c0645-1681-4874-ba8c-37e1bb19262d
* title: Review Clarifying Questions
* topics: pse
##### !question

While we build our skills in breaking down a problem and choosing clarifying questions, let’s use an external tool like ChatGPT to review the questions we wrote above. Our goals are to: 
- ensure we are asking questions that will tell us new information about the problem space
- check our understanding of the information we expect to get from those questions
- uncover other questions that could be useful to ask
- understand why those other questions could be helpful

<br>

For this question we will:

> 1. Build a prompt using [the template linked here](../shared-resources/questions-review-prompt-template.md)

Author: I'd love to get other folk's perspective on this. We can distribute the templates, but then would we not link anything here and have a line telling them to go get it at X place? Even when we provide a resource elsewhere, our process is generally to link a resource where it is used for ease of access. 

I know it is deeply disappointing to lose work, and also, I think students should extremely quickly be very comfortable opening things in another tab.

2. Share the prompt with an AI tool like ChatGPT
3. Hold a discussion with the tool to ensure our understanding of what new information we would get from the initial questions submitted, and what other questions we might want to ask.
4. Reflect on the information shared by the AI tool and summarize its findings

<br>

In the box below, please submit:
1. A shareable link to your conversation in ChatGPT 

>     - [Documentation for creating a shareable link in ChatGPT](https://help.openai.com/en/articles/7925741-chatgpt-shared-links-faq)

Reviewer 3: What if we included some potential follow-up questions, like "How did you select on what to base your analysis of my questions?" Like I'm wondering whether there's a way we can get Chat to spit out more general principles that students could then apply to other PSEs. There's _some_ of that in the "why ask this" and "what does it clarify" bits it tacks on, but can we model how students can extract that more general knowledge directly?

>     - [Documentation for creating a shareable link in ChatGPT](https://help.openai.com/en/articles/7925741-chatgpt-shared-links-faq)

Author: I think it would be really helpful to add some follow up questions to better show how it could be a conversation and how to dig into the understanding deeper.

    - If so, what updates and why?
- Did ChatGPT suggest any new questions?
    - If so, what questions and why would they be useful?

##### !end-hint
##### !explanation 

Everyone’s questions and conversation with ChatGPT will be a little different, and thus the summaries will look a little different. As an example, let’s say we used the questions and explanations below to create our prompt:

```text
1. Does capitalization matter?
    - I want to know whether we need to tell the difference between “Rock” and “rock” 
2. Does extra whitespace / punctuation matter?
    - I want to know whether inputs like " rock " or "rock!" are valid or should be rejected. 
3. Should anything be printed to the console?
    - I know that we want to return the result of the match based on the problem statement, but should we also print the result for the user? 
4. Should there be a default value for each of the arguments? 
    - I want to know whether the function is intended to be flexible and forgiving with missing inputs, or if both arguments are strictly required.
```

<br>

Depending on exactly what ChatGPT shares, a reflection and summary might look like:

<br>

> Chat link: https://chatgpt.com/share/682fa17f-b1a8-8005-9dec-301093934696 

Author: Something like an internal .md link would be more reliable. One of the things I wanted to ask the team is if we want to include links to a sample conversation in our sample response or just include a placeholder like `<url to your chat>`.

    # assert
    assert result == "Player 1 wins!"

def test_winner_both_rock_results_in_tie():
    # arrange
    player_1 = "rock"
    player_2 = "rock"

    # act
    result = winner(player_1, player_2)

    # assert
    assert result == "It's a tie!"  
```

##### !end-explanation
### !end-challenge
<!-- prettier-ignore-end -->

<!-- prettier-ignore-start -->
### !challenge
* type: paragraph
* id: 2bf3b413-e220-46aa-a99a-6c5f3d9715a9
* title: Create Logical Steps
* topics: pse
##### !question

> Without writing code, describe how you would implement `winner` in enough detail that another Python developer could reasonably implement a solution. We should capture the main use cases, but the steps do not need to be a detailed plan for every contingency. 

Reviewer: Repeated comment from `highest-rated.md`:
```suggestion
Without writing code, describe how you would implement `winner` in enough detail that another developer could reasonably implement a solution. We should capture the main use cases, but the steps do not need to be a detailed plan for every contingency. 
```

### !end-challenge
<!-- prettier-ignore-end -->

<!-- prettier-ignore-start -->
### !challenge
* type: paragraph
* id: 2bf3b413-e220-46aa-a99a-6c5f3d9715a9
* title: Create Logical Steps
* topics: pse
##### !question

> Without writing code, describe how you would implement `winner` in enough detail that someone else could write the code. 

Author: I agree that's an easy place to get caught up and think you need to account for each line. What about something like:

```suggestion
Without writing code, describe how you would implement `winner` in enough detail that an experienced developer could reasonably write an implementation based on the description. We should capture the main use cases, but the steps do not need to be a detailed plan for every contingency. The objective is to create a roadmap that we can use to keep ourselves oriented towards our goal, it is okay to leave some of the finer details to be worked out in the implementation itself!
```

    - otherwise player_2 must be "paper", return "Player 2 wins!"
3. Next deal with the case that player_1 is "paper"
    - if player_2 is "rock", return "Player 1 wins!"
    - otherwise player_2 must be "scissors", return "Player 2 wins!"
4. Otherwise player_1 must be "scissors"
    - if player_2 is "rock", return "Player 2 wins!"
    - otherwise player_2 must be "paper", return "Player 1 wins!"

##### !end-explanation
### !end-challenge
<!-- prettier-ignore-end -->

<!-- prettier-ignore-start -->
### !challenge
* type: paragraph
* id: eb41ea7a-cbdf-46d7-b48d-a03bebe872f7
* title: Review Logical Steps
* topics: pse
##### !question

We want to know if we are laying out an approach to the coding problem that makes sense for our context and if that approach is clearly conveying our thoughts on technical topics to others. Let’s once more use an AI tool like ChatGPT, this time to review the Logical Steps we wrote above. Our goals are to check if:
- the steps make sense for the problem being solved
- the steps are not missing important steps or scenarios
- the steps are agnostic of any particular language – steps should not include code syntax.

> - the steps are written with enough detail for another Python developer to understand how to create a solution

Author: ```suggestion
- the steps are written with enough detail for another developer to understand how to create a solution
```

<br>

For this question we will:
1. Build a prompt using [the template linked here](https://gist.githubusercontent.com/ada-instructors/670252696f1625cf0ed77c0997cd165d/raw/pse_logical_steps_review_template.md)

> 2. Share the filled in prompt with an AI tool like ChatGPT

Reviewer: ```suggestion
2. Share the completed prompt with an AI tool like ChatGPT
```

    # assert
    assert result == "Player 1 wins!"

def test_winner_both_rock_returns_tie():
    # arrange
    player_1 = "rock"

>     player_2 = "rock"

Reviewer: Thank you for taking this test case out!
